### crisp PCM の生成

- $n$: 基準/代替案数
- $V$: 一対比較値として取りうる値の集合 ($\{ 1/7, 1/5, 1/3, 1, 3, 5, 7 \}$ など)

1. $i < j$ について $a_{ij}$ を $V$ の中からランダムに選ぶ
2. $i > j$ について $a_{ij} = 1 / a_{ji}$ とする

### 区間 PCM の生成

- $n$: 基準/代替案数
- $N = \left\{ 1, 2, \dots, n \right\}$
- $r_{\max}$: 区間 PCM $A_k$ の各成分の幅を定めるときのパラメータ

crisp PCM $A = \left( a_{ij} \right)_{n \times n}$ を与える.

$\mathrm{DM}_k$ の区間 PCM $A_k = \left( A_{kij} \right)_{n \times n} = \left( \left[ a_{kij}^\mathrm{L}, a_{kij}^\mathrm{U} \right] \right)_{n \times n}$ を次のように与える.

1. $i < j$ の成分を次のように計算する.
    $$
    \begin{align*}
    a_{kij}^\mathrm{L} & = a_{ij} \exp \left( -r_{kij}^\mathrm{L} \right), \quad
    a_{kij}^\mathrm{U} & = a_{ij} \exp \left( r_{kij}^\mathrm{U} \right), \quad
    i, j \in N, ~~ i < j.
    \end{align*}
    $$
    ただし, $r_{kij}^\mathrm{L}, r_{kij}^\mathrm{U} ~ (k \in \{ 1,2 \}, ~~ i, j \in N, ~ i < j)$ は $0$ から $r_{\max}$ の一様分布に従う.

2. $i > j$ の成分は $A_{kij} = 1 / A_{kji} = \left[ \frac{1}{a_{kij}^\mathrm{U}}, \frac{1}{a_{kij}^\mathrm{L}} \right]$ とする.

### シミュレーション

1. 整合している crisp PCM $A_1, A_2$ を 1000 組生成する
2. crisp PCM 1 個に対して, 区間 PCM を 10 個生成する
3. 各区間 PCM について
    1. 手法 1 ~ 4 によって二重区間 PCM $\mathscr{A}^1, \mathscr{A}^2, \mathscr{A}^3, \mathscr{A}^4$ を計算
    2. 類似度を計算
        $$
        \mathrm{Sim}\left( A_{kij}, \mathscr{A}_{ij}^{l\pm} \right), ~~ i, j \in N, ~ i \neq j, ~ k \in \{ 1, 2 \}, ~ l \in \{ 1, 2, 3, 4 \}
        $$
    3. 類似度の集計
        1. $i,j$ 成分の算術平均, 二人の算術平均をとる
            $$
            \rho^{l\pm} = \frac{1}{2n(n-1)} \sum_{k \in \{ 1,2 \}} \sum_{i, j \in N, i \neq j} \mathrm{Sim}\left( A_{kij}, \mathscr{A}_{ij}^{l\pm} \right), ~~ l \in \{ 1, 2, 3, 4 \}
            $$
        2. $i,j$ 成分の算術平均, 二人の幾何平均をとる
            $$
            \sigma^{l\pm} = \sqrt{ \frac{1}{n(n-1)} \prod_{k \in \{ 1,2 \}} \sum_{i, j \in N, i \neq j} \mathrm{Sim}\left( A_{kij}, \mathscr{A}_{ij}^{l\pm} \right) }, ~~ l \in \{ 1, 2, 3, 4 \}
            $$
        <!-- 3. それぞれの手法の算術平均について, 二人の差をとる
            $$
            d^{l\pm} = \left| \sum_{i, j \in N, i \neq j}  \left( \mathrm{Sim}\left( A_{1ij}, \mathscr{A}_{ij}^{l\pm} \right) - \mathrm{Sim}\left( A_{2ij}, \mathscr{A}_{ij}^{l\pm} \right) \right) \right| , ~~ l \in \{ 1, 2, 3, 4 \}
            $$ -->
3. $\rho^{l\pm}, \sigma^{l\pm}, ~ l \in \{ 1, 2, 3, 4 \}$ がそれぞれ 10000 個得られるので, その算術平均をとる

### 結果

PCM のサイズ別に表にしている.  
行が区間の幅に関する変数 $r_{\max}$ で, 下の行ほど幅が大きくなりうる.  
左 4 列が二重区間の内側で, 右 4 列が外側.

### $\rho^{l\pm}$

$n = 5$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0012|0.0000|0.0000|0.0000|0.0719|0.1531|0.1366|0.1677|
|0.5|0.0096|0.0000|0.0008|0.0002|0.2758|0.3452|0.2726|0.3718|
|1.0|0.0652|0.0032|0.0624|0.0194|0.4165|0.4156|0.3893|0.4727|
|1.6|0.1952|0.0322|0.2811|0.1651|0.4986|0.4340|0.4767|0.5181|

$n = 6$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0461|0.0902|0.0889|0.1139|
|0.5|0.0007|0.0000|0.0008|0.0002|0.1929|0.2596|0.1931|0.2833|
|1.0|0.0089|0.0000|0.0005|0.0001|0.3132|0.3359|0.2745|0.3859|
|1.6|0.0546|0.0004|0.0418|0.0053|0.3937|0.3631|0.4767|0.4393|

$n = 7$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0344|0.0606|0.0613|0.0806|
|0.5|0.0000|0.0000|0.0000|0.0000|0.1482|0.2021|0.1692|0.2282|
|1.0|0.0009|0.0000|0.0000|0.0000|0.2485|0.2808|0.2559|0.3289|
|1.6|0.0095|0.0000|0.0006|0.0000|0.3246|0.3125|0.3222|0.3876|

$n = 8$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0286|0.0483|0.0425|0.0616|
|0.5|0.0000|0.0000|0.0000|0.0000|0.1240|0.1710|0.1389|0.1981|
|1.0|0.0001|0.0000|0.0000|0.0000|0.2117|0.2454|0.2246|0.2963|
|1.6|0.0013|0.0000|0.0000|0.0000|0.2827|0.2775|0.2892|0.3532|

### $\sigma^{l\pm}$

$n = 5$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0006|0.0|0.0|0.0|0.0357|0.0759|0.0676|0.0831|
|0.5|0.0046|0.0001|0.0006|0.0003|0.1373|0.1721|0.1367|0.1853|
|1.0|0.0307|0.0009|0.0293|0.0072|0.2075|0.2070|0.1922|0.2355|
|1.6|0.0933|0.0143|0.1359|0.0787|0.2464|0.2152|0.2352|0.2571|

$n = 6$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0231|0.0455|0.0438|0.0571|
|0.5|0.0003|0.0000|0.0000|0.0000|0.0961|0.1294|0.0961|0.1413|
|1.0|0.0046|0.0000|0.0005|0.0001|0.1569|0.1687|0.1382|0.1942|
|1.6|0.0267|0.0001|0.0189|0.0026|0.1956|0.1813|0.1801|0.2198|

$n = 7$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0171|0.0299|0.0301|0.0398|
|0.5|0.0000|0.0000|0.0000|0.0000|0.0741|0.1012|0.0858|0.1146|
|1.0|0.0004|0.0000|0.0000|0.0000|0.1239|0.1394|0.1287|0.1643|
|1.6|0.0046|0.0000|0.0004|0.0000|0.1617|0.1562|0.1609|0.1932|

$n = 8$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0142|0.0242|0.0213|0.0312|
|0.5|0.0000|0.0000|0.0000|0.0000|0.0616|0.0858|0.0701|0.0994|
|1.0|0.0000|0.0000|0.0000|0.0000|0.1057|0.1219|0.1110|0.1473|
|1.6|0.0006|0.0000|0.0000|0.0000|0.1406|0.1383|0.1449|0.1764|


### 考察

- $n$ と一致度の関係
    - いずれのケースでも $n$ が大きくなると一致度が下がる
        - 制約が増え, 整合化が行われることがより多くなるため
- $r_{\max}$ と一致度の関係
    - いずれのケースでも $r_{\max}$ が大きくなると一致度が上がる
        - 区間が広くなると整合化が減る
- 内側の区間の一致度 $\rho^{l-}$ の手法間の比較
    - $n \geq 6$ では一致度 $0$ がほとんど
        - 区間の幅を広げたシミュレーションなどが必要 
    - 手法 1 がほとんどの場合最も良い
    - $n = 5, r_{\max} = 1.6$ では手法 3 が最も良く, 差があって手法 1, 4 と続く
        - これまでの数値例はこのケースに該当する
    - 区間の幅を広げていく ($r_{\max}$ を大きくしていく) と個人の重要度を先に求める手法 3, 4 の一致度が大きく改善する
- 外側の区間の一致度 $\rho^{l+}$ の手法間の比較
    - ほとんどのケースで一致度が手法  4 が最も良い
        - $n = 6, r_{\max} = 1.6$ のケースのみ手法 6 のほうが良い

---

In [1]:
println(Threads.nthreads())

24


In [2]:
using IntervalArithmetic
using Statistics

In [3]:
include("./crispPCM/index.jl")
include("./evaluation/index.jl")
include("./method1/index.jl")
include("./method2/index.jl")
include("./method3/index.jl")
include("./method4/index.jl")
include("./intervalPCM/index.jl")
include("./twofoldInterval/index.jl")
include("./twofoldIntervalPCM/index.jl")
include("./utils.jl")

R (generic function with 1 method)

In [4]:
Random.seed!(1)

TaskLocalRNG()

In [5]:
SimulationCase = @NamedTuple{
    CR::T,
    A::Matrix{T},
    A₁::Matrix{Interval{T}},
    A₂::Matrix{Interval{T}}
} where {T <: Real}

function generateSimulationCases(
    n::Integer, 
    values::Vector{T},
    width::T,
    numOfCrispPCM::Integer, 
    intervalPCMsPerCrispPCM::Integer
)::Array{SimulationCase, 1} where {T <: Real}

    simulation_cases = SimulationCase[]

    for _ in 1:numOfCrispPCM
        # Generate a Crisp PCM
        crisp_pcm = generateConsistentCrispPCM(n, values)
        CR = consistencyRatio(crisp_pcm)

        for _ in 1:intervalPCMsPerCrispPCM
            # Generate an Interval PCM for each Crisp PCM
            interval_pcm_1 = randamizedIntervalPCM(crisp_pcm, width)
            interval_pcm_2 = randamizedIntervalPCM(crisp_pcm, width)
            # Append the Crisp PCM and its corresponding Interval PCM as a tuple to the list
            push!(simulation_cases, (CR=CR, A=crisp_pcm, A₁=interval_pcm_1, A₂=interval_pcm_2))
        end
    end

    return simulation_cases
end


generateSimulationCases (generic function with 1 method)

In [6]:
SimulationResult = @NamedTuple{
    # A₁, A₂ の元となった crisp PCM
    A::Matrix{T},
    # 各 DM の与える区間 PCM
    A₁::Matrix{Interval{T}}, A₂::Matrix{Interval{T}},
    # method 1 ~ 4 の二重区間 PCM
    𝓐¹::Matrix{TwofoldInterval{T}}, 𝓐²::Matrix{TwofoldInterval{T}},
    𝓐³::Matrix{TwofoldInterval{T}}, 𝓐⁴::Matrix{TwofoldInterval{T}},
    # 𝓐ᵏ の二重区間の内側の区間
    𝓐¹⁻::Matrix{Interval{T}}, 𝓐²⁻::Matrix{Interval{T}},
    𝓐³⁻::Matrix{Interval{T}}, 𝓐⁴⁻::Matrix{Interval{T}},
    # 𝓐ᵏ の二重区間の外側の区間
    𝓐¹⁺::Matrix{Interval{T}}, 𝓐²⁺::Matrix{Interval{T}},
    𝓐³⁺::Matrix{Interval{T}}, 𝓐⁴⁺::Matrix{Interval{T}},
    # エラーメッセージ
    error::String
} where {T <: Real}

methodList = [method1, method2, method3, method4]

function runSimulation(
    case::SimulationCase{T}
    )::SimulationResult{T} where {T <: Real}
    𝓐 = Dict(); 𝓐⁻ = Dict(); 𝓐⁺ = Dict()
    
    try
        for k in 1:4
            𝓐[k] = methodList[k](case.A₁, case.A₂)
            𝓐⁻[k] = map(𝓐ᵢⱼ -> 𝓐ᵢⱼ[1], 𝓐[k])
            𝓐⁺[k] = map(𝓐ᵢⱼ -> 𝓐ᵢⱼ[2], 𝓐[k])
        end

        return (
            A=case.A, A₁=case.A₁, A₂=case.A₂,
            𝓐¹=𝓐[1], 𝓐²=𝓐[2], 𝓐³=𝓐[3], 𝓐⁴=𝓐[4],
            𝓐¹⁻=𝓐⁻[1], 𝓐²⁻=𝓐⁻[2], 𝓐³⁻=𝓐⁻[3], 𝓐⁴⁻=𝓐⁻[4],
            𝓐¹⁺=𝓐⁺[1], 𝓐²⁺=𝓐⁺[2], 𝓐³⁺=𝓐⁺[3], 𝓐⁴⁺=𝓐⁺[4],
            error=""
        )
    catch e
        n = size(case.A₁, 2)
        IPCM = fill(1..1, n, n)
        twofoldIPCM = fill((1..1, 1..1), n, n)
        return (
            A=case.A, A₁=case.A₁, A₂=case.A₂,
            𝓐¹=twofoldIPCM, 𝓐²=twofoldIPCM, 𝓐³=twofoldIPCM, 𝓐⁴=twofoldIPCM,
            𝓐¹⁻=IPCM, 𝓐²⁻=IPCM, 𝓐³⁻=IPCM, 𝓐⁴⁻=IPCM,
            𝓐¹⁺=IPCM, 𝓐²⁺=IPCM, 𝓐³⁺=IPCM, 𝓐⁴⁺=IPCM,
            error=string(e)
        )
    end
end

runSimulation (generic function with 1 method)

In [7]:
ConincidenceList = @NamedTuple{
    A₁_𝓐¹⁻::T, A₁_𝓐¹⁺::T, A₂_𝓐¹⁻::T, A₂_𝓐¹⁺::T,
    A₁_𝓐²⁻::T, A₁_𝓐²⁺::T, A₂_𝓐²⁻::T, A₂_𝓐²⁺::T,
    A₁_𝓐³⁻::T, A₁_𝓐³⁺::T, A₂_𝓐³⁻::T, A₂_𝓐³⁺::T,
    A₁_𝓐⁴⁻::T, A₁_𝓐⁴⁺::T, A₂_𝓐⁴⁻::T, A₂_𝓐⁴⁺::T,
    error::String
} where {T <: Real}

function calculateCoincidenceList(result::SimulationResult{T})::ConincidenceList{T} where {T <: Real}
    A₁ = result.A₁; A₂ = result.A₂
    𝓐¹⁻ = result.𝓐¹⁻; 𝓐¹⁺ = result.𝓐¹⁺
    𝓐²⁻ = result.𝓐²⁻; 𝓐²⁺ = result.𝓐²⁺
    𝓐³⁻ = result.𝓐³⁻; 𝓐³⁺ = result.𝓐³⁺
    𝓐⁴⁻ = result.𝓐⁴⁻; 𝓐⁴⁺ = result.𝓐⁴⁺
    try
        return (
            A₁_𝓐¹⁻=coincidenceIndex(A₁, 𝓐¹⁻), A₁_𝓐¹⁺=coincidenceIndex(A₁, 𝓐¹⁺),
            A₂_𝓐¹⁻=coincidenceIndex(A₂, 𝓐¹⁻), A₂_𝓐¹⁺=coincidenceIndex(A₂, 𝓐¹⁺),
            A₁_𝓐²⁻=coincidenceIndex(A₁, 𝓐²⁻), A₁_𝓐²⁺=coincidenceIndex(A₁, 𝓐²⁺),
            A₂_𝓐²⁻=coincidenceIndex(A₂, 𝓐²⁻), A₂_𝓐²⁺=coincidenceIndex(A₂, 𝓐²⁺),
            A₁_𝓐³⁻=coincidenceIndex(A₁, 𝓐³⁻), A₁_𝓐³⁺=coincidenceIndex(A₁, 𝓐³⁺),
            A₂_𝓐³⁻=coincidenceIndex(A₂, 𝓐³⁻), A₂_𝓐³⁺=coincidenceIndex(A₂, 𝓐³⁺),
            A₁_𝓐⁴⁻=coincidenceIndex(A₁, 𝓐⁴⁻), A₁_𝓐⁴⁺=coincidenceIndex(A₁, 𝓐⁴⁺),
            A₂_𝓐⁴⁻=coincidenceIndex(A₂, 𝓐⁴⁻), A₂_𝓐⁴⁺=coincidenceIndex(A₂, 𝓐⁴⁺),
            error=""
        )
    catch e
        return (
            A₁_𝓐¹⁻=NaN, A₁_𝓐¹⁺=NaN, A₂_𝓐¹⁻=NaN, A₂_𝓐¹⁺=NaN,
            A₁_𝓐²⁻=NaN, A₁_𝓐²⁺=NaN, A₂_𝓐²⁻=NaN, A₂_𝓐²⁺=NaN,
            A₁_𝓐³⁻=NaN, A₁_𝓐³⁺=NaN, A₂_𝓐³⁻=NaN, A₂_𝓐³⁺=NaN,
            A₁_𝓐⁴⁻=NaN, A₁_𝓐⁴⁺=NaN, A₂_𝓐⁴⁻=NaN, A₂_𝓐⁴⁺=NaN,
            error=string(e)
        )
    end
end

calculateCoincidenceList (generic function with 1 method)

In [8]:
function averageConincidenceLists(conincidenceLists::Vector{ConincidenceList{T}}) where {T <: Real}
    if length(conincidenceLists) == 0
        throw(ArgumentError("Empty list of ConincidenceLists"))
    end

    total_𝓐¹⁻ = zero(T); total_𝓐¹⁺ = zero(T)
    total_𝓐²⁻ = zero(T); total_𝓐²⁺ = zero(T)
    total_𝓐³⁻ = zero(T); total_𝓐³⁺ = zero(T)
    total_𝓐⁴⁻ = zero(T); total_𝓐⁴⁺ = zero(T)

    total_gm_𝓐¹⁻ = zero(T); total_gm_𝓐¹⁺ = zero(T)
    total_gm_𝓐²⁻ = zero(T); total_gm_𝓐²⁺ = zero(T)
    total_gm_𝓐³⁻ = zero(T); total_gm_𝓐³⁺ = zero(T)
    total_gm_𝓐⁴⁻ = zero(T); total_gm_𝓐⁴⁺ = zero(T)

    total_df_𝓐¹⁻ = zero(T); total_df_𝓐¹⁺ = zero(T)
    total_df_𝓐²⁻ = zero(T); total_df_𝓐²⁺ = zero(T)
    total_df_𝓐³⁻ = zero(T); total_df_𝓐³⁺ = zero(T)
    total_df_𝓐⁴⁻ = zero(T); total_df_𝓐⁴⁺ = zero(T)

    count = 0

    for list in conincidenceLists
        if list.error != "" continue end

        count += 1
        total_𝓐¹⁻ += list.A₁_𝓐¹⁻ + list.A₂_𝓐¹⁻
        total_𝓐¹⁺ += list.A₁_𝓐¹⁺ + list.A₂_𝓐¹⁺
        total_𝓐²⁻ += list.A₁_𝓐²⁻ + list.A₂_𝓐²⁻
        total_𝓐²⁺ += list.A₁_𝓐²⁺ + list.A₂_𝓐²⁺
        total_𝓐³⁻ += list.A₁_𝓐³⁻ + list.A₂_𝓐³⁻
        total_𝓐³⁺ += list.A₁_𝓐³⁺ + list.A₂_𝓐³⁺
        total_𝓐⁴⁻ += list.A₁_𝓐⁴⁻ + list.A₂_𝓐⁴⁻
        total_𝓐⁴⁺ += list.A₁_𝓐⁴⁺ + list.A₂_𝓐⁴⁺

        total_gm_𝓐¹⁻ += sqrt(list.A₁_𝓐¹⁻ * list.A₂_𝓐¹⁻)
        total_gm_𝓐¹⁺ += sqrt(list.A₁_𝓐¹⁺ * list.A₂_𝓐¹⁺)
        total_gm_𝓐²⁻ += sqrt(list.A₁_𝓐²⁻ * list.A₂_𝓐²⁻)
        total_gm_𝓐²⁺ += sqrt(list.A₁_𝓐²⁺ * list.A₂_𝓐²⁺)
        total_gm_𝓐³⁻ += sqrt(list.A₁_𝓐³⁻ * list.A₂_𝓐³⁻)
        total_gm_𝓐³⁺ += sqrt(list.A₁_𝓐³⁺ * list.A₂_𝓐³⁺)
        total_gm_𝓐⁴⁻ += sqrt(list.A₁_𝓐⁴⁻ * list.A₂_𝓐⁴⁻)
        total_gm_𝓐⁴⁺ += sqrt(list.A₁_𝓐⁴⁺ * list.A₂_𝓐⁴⁺)

        total_df_𝓐¹⁻ += abs(list.A₁_𝓐¹⁻ - list.A₂_𝓐¹⁻)
        total_df_𝓐¹⁺ += abs(list.A₁_𝓐¹⁺ - list.A₂_𝓐¹⁺)
        total_df_𝓐²⁻ += abs(list.A₁_𝓐²⁻ - list.A₂_𝓐²⁻)
        total_df_𝓐²⁺ += abs(list.A₁_𝓐²⁺ - list.A₂_𝓐²⁺)
        total_df_𝓐³⁻ += abs(list.A₁_𝓐³⁻ - list.A₂_𝓐³⁻)
        total_df_𝓐³⁺ += abs(list.A₁_𝓐³⁺ - list.A₂_𝓐³⁺)
        total_df_𝓐⁴⁻ += abs(list.A₁_𝓐⁴⁻ - list.A₂_𝓐⁴⁻)
        total_df_𝓐⁴⁺ += abs(list.A₁_𝓐⁴⁺ - list.A₂_𝓐⁴⁺)
    end
    count *= 2

    return (
        ρ¹⁻ = total_𝓐¹⁻ / count, ρ¹⁺ = total_𝓐¹⁺ / count,
        ρ²⁻ = total_𝓐²⁻ / count, ρ²⁺ = total_𝓐²⁺ / count,
        ρ³⁻ = total_𝓐³⁻ / count, ρ³⁺ = total_𝓐³⁺ / count,
        ρ⁴⁻ = total_𝓐⁴⁻ / count, ρ⁴⁺ = total_𝓐⁴⁺ / count,
        σ¹⁻ = total_gm_𝓐¹⁻ / count, σ¹⁺ = total_gm_𝓐¹⁺ / count,
        σ²⁻ = total_gm_𝓐²⁻ / count, σ²⁺ = total_gm_𝓐²⁺ / count,
        σ³⁻ = total_gm_𝓐³⁻ / count, σ³⁺ = total_gm_𝓐³⁺ / count,
        σ⁴⁻ = total_gm_𝓐⁴⁻ / count, σ⁴⁺ = total_gm_𝓐⁴⁺ / count,
        d¹⁻ = total_df_𝓐¹⁻ / count, d¹⁺ = total_df_𝓐¹⁺ / count,   
        d²⁻ = total_df_𝓐²⁻ / count, d²⁺ = total_df_𝓐²⁺ / count,
        d³⁻ = total_df_𝓐³⁻ / count, d³⁺ = total_df_𝓐³⁺ / count,
        d⁴⁻ = total_df_𝓐⁴⁻ / count, d⁴⁺ = total_df_𝓐⁴⁺ / count
    )
end

averageConincidenceLists (generic function with 1 method)

In [9]:
function runSimulate(
    n::Integer, 
    values::Vector{T},
    width::T,
    numOfCrispPCM::Integer, 
    intervalPCMsPerCrispPCM::Integer
    ) where {T <: Real}
    cases = generateSimulationCases(n, values, width, numOfCrispPCM, intervalPCMsPerCrispPCM)
    results = Vector{SimulationResult}(undef, length(cases))
    Threads.@threads for i in 1:length(cases)
        results[i] = runSimulation(cases[i])
    end
    conincidenceList = calculateCoincidenceList.(results)
    return averageConincidenceLists(conincidenceList)
end

runSimulate (generic function with 1 method)

In [10]:
values = [0.1, 0.5, 1.0, 1.6]

for n in 5:8
    println("n = $n")
    for value in values
        println(runSimulate(n, [1/9, 1/7, 1/5, 1/3, 1, 3, 5, 7, 9], value, 1000, 10))
    end
end

n = 5


(ρ¹⁻ = 0.0012292655982896199, ρ¹⁺ = 0.07187403867569402, ρ²⁻ = 4.537924495082146e-16, ρ²⁺ = 0.1531243873253255, ρ³⁻ = 0.0, ρ³⁺ = 0.1366362230294785, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.16767902615045777, σ¹⁻ = 0.0005914130936857436, σ¹⁺ = 0.03574744480008097, σ²⁻ = 2.0109178467630937e-16, σ²⁺ = 0.07593350954220626, σ³⁻ = 0.0, σ³⁺ = 0.06756871791157593, σ⁴⁻ = 0.0, σ⁴⁺ = 0.08314056411167918, d¹⁻ = 0.00026675563121265693, d¹⁺ = 0.00582084985369821, d²⁻ = 1.4792511470529866e-16, d²⁺ = 0.015462939478269965, d³⁻ = 0.0, d³⁺ = 0.015484596231881309, d⁴⁻ = 0.0, d⁴⁺ = 0.016856993517707693)


(ρ¹⁻ = 0.009532676047360451, ρ¹⁺ = 0.27595601187524776, ρ²⁻ = 0.0001567899385967774, ρ²⁺ = 0.34600385264924804, ρ³⁻ = 0.0011692785418443178, ρ³⁺ = 0.27518977837452774, ρ⁴⁻ = 0.0006059074980231841, ρ⁴⁺ = 0.3720651340946489, σ¹⁻ = 0.004598787966682697, σ¹⁺ = 0.13733475403517253, σ²⁻ = 7.649350300677591e-5, σ²⁺ = 0.1720888721323408, σ³⁻ = 0.0005788371236285056, σ³⁺ = 0.1366689676263204, σ⁴⁻ = 0.00030251134582917914, σ⁴⁺ = 0.18527880523054702, d¹⁻ = 0.0018898915619819717, d¹⁺ = 0.02111954061315004, d²⁻ = 3.142875533351838e-5, d²⁺ = 0.028136318020358864, d³⁻ = 0.00011809949905428196, d³⁺ = 0.024855191001163093, d⁴⁻ = 2.9789687988395426e-5, d⁴⁺ = 0.026686095187481537)


(ρ¹⁻ = 0.06274662900199893, ρ¹⁺ = 0.41700462254692905, ρ²⁻ = 0.0017780661356284284, ρ²⁺ = 0.4162274662331469, ρ³⁻ = 0.05961220792093334, ρ³⁺ = 0.38696179005954423, ρ⁴⁻ = 0.014389745786975452, ρ⁴⁺ = 0.47279998387600164, σ¹⁻ = 0.03068071952481451, σ¹⁺ = 0.20746479050569713, σ²⁻ = 0.0008764677906932903, σ²⁺ = 0.20698633696803714, σ³⁻ = 0.02933752665299981, σ³⁺ = 0.19221969947721135, σ⁴⁻ = 0.007153753618123909, σ⁴⁺ = 0.2355318734354983, d¹⁻ = 0.009678171103967854, d¹⁺ = 0.03307124754222082, d²⁻ = 0.0002304456260578315, d²⁺ = 0.034387625261472585, d³⁻ = 0.008151150038468974, d³⁺ = 0.03491762894753613, d⁴⁻ = 0.0012399987286105135, d⁴⁺ = 0.032143005033992526)


(ρ¹⁻ = 0.18907008464729883, ρ¹⁺ = 0.49540556711674416, ρ²⁻ = 0.02901091937570912, ρ²⁺ = 0.43305322921695977, ρ³⁻ = 0.27474486324390507, ρ³⁺ = 0.4734772744820122, ρ⁴⁻ = 0.1582714983533993, ρ⁴⁺ = 0.5160804511158598, σ¹⁻ = 0.09333104581659404, σ¹⁺ = 0.24635566838077005, σ²⁻ = 0.01434633634362714, σ²⁺ = 0.21519657245954776, σ³⁻ = 0.13588458685190982, σ³⁺ = 0.23516123233962588, σ⁴⁻ = 0.07865521103356753, σ⁴⁺ = 0.25711105016876556, d¹⁻ = 0.022976909523000744, d¹⁺ = 0.041093072721539926, d²⁻ = 0.0033498656600487757, d²⁺ = 0.03812069921249145, d³⁻ = 0.031143336144119814, d³⁺ = 0.04369032262640971, d⁴⁻ = 0.013926647436258096, d⁴⁺ = 0.03489141654920768)
n = 6


(ρ¹⁻ = 3.8269040171759984e-5, ρ¹⁺ = 0.046440889740986674, ρ²⁻ = 2.978834308245805e-16, ρ²⁺ = 0.09142678914369774, ρ³⁻ = 0.0, ρ³⁺ = 0.0884844108610448, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.11516368202398908, σ¹⁻ = 1.85282054099663e-5, σ¹⁺ = 0.023145966158814982, σ²⁻ = 1.3316905938470572e-16, σ²⁺ = 0.04545855634066476, σ³⁻ = 0.0, σ³⁺ = 0.043761137805796095, σ⁴⁻ = 0.0, σ⁴⁺ = 0.05714851411962113, d¹⁻ = 7.559811197279707e-6, d¹⁺ = 0.0029570261226166536, d²⁻ = 9.288659405715966e-17, d²⁺ = 0.0075490745109442646, d³⁻ = 0.0, d³⁺ = 0.009984345488074057, d⁴⁻ = 0.0, d⁴⁺ = 0.01101382121113853)


(ρ¹⁻ = 0.0007082950049671125, ρ¹⁺ = 0.1928357868909078, ρ²⁻ = 9.968725163726371e-17, ρ²⁺ = 0.2596941053103204, ρ³⁻ = 0.0, ρ³⁺ = 0.19329841991014055, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.28339705349412503, σ¹⁻ = 0.000337835657510278, σ¹⁺ = 0.09611175443589484, σ²⁻ = 4.704649333061432e-17, σ²⁺ = 0.12935779651324714, σ³⁻ = 0.0, σ³⁺ = 0.09605328270859576, σ⁴⁻ = 0.0, σ⁴⁺ = 0.14126768387931046, d¹⁻ = 0.00016248267156440547, d¹⁺ = 0.012199158440778525, d²⁻ = 2.4286627566980857e-17, d²⁺ = 0.017904676999738228, d³⁻ = 0.0, d³⁺ = 0.016533319304101242, d⁴⁻ = 0.0, d⁴⁺ = 0.01759647999230751)


(ρ¹⁻ = 0.009562480600161431, ρ¹⁺ = 0.31479768918210255, ρ²⁻ = 4.1465538004028165e-5, ρ²⁺ = 0.33869231352005635, ρ³⁻ = 0.0010521820274404568, ρ³⁺ = 0.2776581402452136, ρ⁴⁻ = 0.00019012856242566124, ρ⁴⁺ = 0.3894017113087499, σ¹⁻ = 0.004629168919222696, σ¹⁺ = 0.15686267464011588, σ²⁻ = 2.069816290044548e-5, σ²⁺ = 0.16870317737073243, σ³⁻ = 0.000521832379898929, σ³⁺ = 0.13817593805579154, σ⁴⁻ = 9.504285652145179e-5, σ⁴⁺ = 0.19419513063565386, d¹⁻ = 0.001761095247849796, d¹⁺ = 0.02069220387429978, d²⁻ = 2.1971583802098525e-6, d²⁺ = 0.02352702340271543, d³⁻ = 0.00010089845117082735, d³⁺ = 0.02094493801508495, d⁴⁻ = 3.2298218138080544e-6, d⁴⁺ = 0.022344940242406616)


(ρ¹⁻ = 0.05443908765434667, ρ¹⁺ = 0.3927516659309165, ρ²⁻ = 0.00023045271513688878, ρ²⁺ = 0.36408491328882286, ρ³⁻ = 0.03824383537933709, ρ³⁺ = 0.36182493259827675, ρ⁴⁻ = 0.005257742993888137, ρ⁴⁺ = 0.44086709698906074, σ¹⁻ = 0.026697240026348098, σ¹⁺ = 0.1956370358732584, σ²⁻ = 0.00011394799840204368, σ²⁺ = 0.18126568279269073, σ³⁻ = 0.018880011833233207, σ³⁺ = 0.18011940649189523, σ⁴⁻ = 0.0026174952531703316, σ⁴⁺ = 0.2198428901428306, d¹⁻ = 0.007960949716567895, d¹⁺ = 0.02712022751062288, d²⁻ = 2.7387659546521067e-5, d²⁺ = 0.026707946290038847, d³⁻ = 0.004817543037722012, d³⁺ = 0.02674389875465382, d⁴⁻ = 0.0003886614203236218, d⁴⁺ = 0.025685030895763506)
n = 7


┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/p

(ρ¹⁻ = 1.2724748571538429e-5, ρ¹⁺ = 0.03438252050517541, ρ²⁻ = 1.9845294685462477e-16, ρ²⁺ = 0.06000647278360889, ρ³⁻ = 0.0, ρ³⁺ = 0.060624939598353945, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.08003974188669144, σ¹⁻ = 6.228219488172145e-6, σ¹⁺ = 0.017148739166719166, σ²⁻ = 8.836120243608663e-17, σ²⁺ = 0.02990942904221349, σ³⁻ = 0.0, σ³⁺ = 0.030051401586224245, σ⁴⁻ = 0.0, σ⁴⁺ = 0.03977488343893432, d¹⁻ = 2.3929953309858993e-6, d¹⁺ = 0.001906247772635547, d²⁻ = 6.255852080194834e-17, d²⁺ = 0.0037467587878987495, d³⁻ = 0.0, d³⁺ = 0.006084351099563782, d⁴⁻ = 0.0, d⁴⁺ = 0.006929644518177351)


(ρ¹⁻ = 5.6785024278734824e-5, ρ¹⁺ = 0.14862464916956034, ρ²⁻ = 7.090158482556592e-17, ρ²⁺ = 0.2028496581324599, ρ³⁻ = 0.0, ρ³⁺ = 0.17224965399013015, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.22981613543455956, σ¹⁻ = 2.6370857468556757e-5, σ¹⁺ = 0.07412908611952779, σ²⁻ = 3.361876678227496e-17, σ²⁺ = 0.1011519704646714, σ³⁻ = 0.0, σ³⁺ = 0.08579369698350552, σ⁴⁻ = 0.0, σ⁴⁺ = 0.11464569933639311, d¹⁻ = 1.4941770271152993e-5, d¹⁺ = 0.00832643940739571, d²⁻ = 1.657232774358658e-17, d²⁺ = 0.011848598987561862, d³⁻ = 0.0, d³⁺ = 0.011803299298596671, d⁴⁻ = 0.0, d⁴⁺ = 0.012374097666620959)


(ρ¹⁻ = 0.0007538658239791082, ρ¹⁺ = 0.24850809114446926, ρ²⁻ = 6.066587708845394e-17, ρ²⁺ = 0.27959745188152524, ρ³⁻ = 0.0, ρ³⁺ = 0.2581521938091992, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.3292488626736071, σ¹⁻ = 0.00036126863854391564, σ¹⁺ = 0.1239286677410143, σ²⁻ = 2.94083157596683e-17, σ²⁺ = 0.13941054732601338, σ³⁻ = 0.0, σ³⁺ = 0.12867546369158808, σ⁴⁻ = 0.0, σ⁴⁺ = 0.1643049357175383, d¹⁻ = 0.00016471918504498887, d¹⁺ = 0.014308283904097706, d²⁻ = 1.1247875676211357e-17, d²⁺ = 0.016597382823660155, d³⁻ = 0.0, d³⁺ = 0.016076398588076378, d⁴⁻ = 0.0, d⁴⁺ = 0.016322105591149976)


(ρ¹⁻ = 0.009449923069612882, ρ¹⁺ = 0.32433308027047225, ρ²⁻ = 5.867981689806825e-17, ρ²⁺ = 0.31333649870571534, ρ³⁻ = 0.0008064185340314108, ρ³⁺ = 0.32268031573363287, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.3872711223593831, σ¹⁻ = 0.004587703408332498, σ¹⁺ = 0.16168181943248916, σ²⁻ = 2.871658842842377e-17, σ²⁺ = 0.15615714170566458, σ³⁻ = 0.0003989927771879872, σ³⁺ = 0.16087348253904055, σ⁴⁻ = 0.0, σ⁴⁺ = 0.1932401059128731, d¹⁻ = 0.0016749053509084808, d¹⁺ = 0.0199617919452534, d²⁻ = 9.065236718721539e-18, d²⁺ = 0.02010094701129514, d³⁻ = 9.534845488988637e-5, d³⁺ = 0.019358896349746305, d⁴⁻ = 0.0, d⁴⁺ = 0.01968628506118533)
n = 8


(ρ¹⁻ = 0.0, ρ¹⁺ = 0.02840630913798917, ρ²⁻ = 1.440579927118098e-16, ρ²⁺ = 0.04847270911140285, ρ³⁻ = 0.0, ρ³⁺ = 0.042891224445762044, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.06270348228109468, σ¹⁻ = 0.0, σ¹⁺ = 0.014175689103804734, σ²⁻ = 6.452405827687369e-17, σ²⁺ = 0.024184033671799363, σ³⁻ = 0.0, σ³⁺ = 0.02129794088762056, σ⁴⁻ = 0.0, σ⁴⁺ = 0.03120834005743195, d¹⁻ = 0.0, d¹⁺ = 0.0014040485878980592, d²⁻ = 4.477656783734837e-17, d²⁺ = 0.002518797572488253, d³⁻ = 0.0, d³⁺ = 0.0037985056411898766, d⁴⁻ = 0.0, d⁴⁺ = 0.004603384301278449)


(ρ¹⁻ = 3.519307968749531e-6, ρ¹⁺ = 0.12339084626558457, ρ²⁻ = 5.470356077486367e-17, ρ²⁺ = 0.17187636253281666, ρ³⁻ = 0.0, ρ³⁺ = 0.1407185228890346, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.19921075280953685, σ¹⁻ = 1.6657983896418737e-6, σ¹⁺ = 0.06157728231332223, σ²⁻ = 2.607609205566357e-17, σ²⁺ = 0.08577355486022649, σ³⁻ = 0.0, σ³⁺ = 0.07014762632140267, σ⁴⁻ = 0.0, σ⁴⁺ = 0.09943735841090012, d¹⁻ = 9.070084904508329e-7, d¹⁺ = 0.006047375147467263, d²⁻ = 1.2270019739673873e-17, d²⁺ = 0.00845556984132353, d³⁻ = 0.0, d³⁺ = 0.008553992148170892, d⁴⁻ = 0.0, d⁴⁺ = 0.009250120293682237)


(ρ¹⁻ = 6.302413336718727e-5, ρ¹⁺ = 0.21184267519051764, ρ²⁻ = 2.9941016674123475e-12, ρ²⁺ = 0.24434996800646194, ρ³⁻ = 0.0, ρ³⁺ = 0.22249107395743675, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.2950870434606053, σ¹⁻ = 3.0295521265476337e-5, σ¹⁺ = 0.10570148955546711, σ²⁻ = 1.118745181859527e-12, σ²⁺ = 0.12191782120532299, σ³⁻ = 0.0, σ³⁺ = 0.11097031650102797, σ⁴⁻ = 0.0, σ⁴⁺ = 0.14732335210133973, d¹⁻ = 1.2801004758255305e-5, d¹⁺ = 0.010882293320006748, d²⁻ = 1.9895316739650368e-12, d²⁺ = 0.012710358769407615, d³⁻ = 0.0, d³⁺ = 0.012271073835068654, d⁴⁻ = 0.0, d⁴⁺ = 0.012822202169077625)


(ρ¹⁻ = 0.0011958522893586673, ρ¹⁺ = 0.2819060138581883, ρ²⁻ = 4.863559233276879e-17, ρ²⁺ = 0.2772243704215146, ρ³⁻ = 0.0, ρ³⁺ = 0.29041534331030217, ρ⁴⁻ = 0.0, ρ⁴⁺ = 0.3532421103279215, σ¹⁻ = 0.0005714021782451148, σ¹⁺ = 0.14061346859287138, σ²⁻ = 2.3848616065785948e-17, σ²⁺ = 0.1382538597870038, σ³⁻ = 0.0, σ³⁺ = 0.14489255793487546, σ⁴⁻ = 0.0, σ⁴⁺ = 0.17635385721794375, d¹⁻ = 0.00025871451057436073, d¹⁺ = 0.015586752896331959, d²⁻ = 7.043528535258468e-18, d²⁺ = 0.015896307872734546, d³⁻ = 0.0, d³⁺ = 0.015167293257391686, d⁴⁻ = 0.0, d⁴⁺ = 0.015460534725261878)


### n = 5

- $r_{\max} = 0.1$

    (σ¹⁻ = 0.0005914130936857436, σ¹⁺ = 0.03574744480008097, σ²⁻ = 2.0109178467630937e-16, σ²⁺ = 0.07593350954220626, σ³⁻ = 0.0, σ³⁺ = 0.06756871791157593, σ⁴⁻ = 0.0, σ⁴⁺ = 0.08314056411167918, d¹⁻ = 0.00026675563121265693, d¹⁺ = 0.00582084985369821, d²⁻ = 1.4792511470529866e-16, d²⁺ = 0.015462939478269965, d³⁻ = 0.0, d³⁺ = 0.015484596231881309, d⁴⁻ = 0.0, d⁴⁺ = 0.016856993517707693)

- $r_{\max} = 0.5$

    (σ¹⁻ = 0.004598787966682697, σ¹⁺ = 0.13733475403517253, σ²⁻ = 7.649350300677591e-5, σ²⁺ = 0.1720888721323408, σ³⁻ = 0.0005788371236285056, σ³⁺ = 0.1366689676263204, σ⁴⁻ = 0.00030251134582917914, σ⁴⁺ = 0.18527880523054702, d¹⁻ = 0.0018898915619819717, d¹⁺ = 0.02111954061315004, d²⁻ = 3.142875533351838e-5, d²⁺ = 0.028136318020358864, d³⁻ = 0.00011809949905428196, d³⁺ = 0.024855191001163093, d⁴⁻ = 2.9789687988395426e-5, d⁴⁺ = 0.026686095187481537)


- $r_{\max} = 1.0$

    (σ¹⁻ = 0.03068071952481451, σ¹⁺ = 0.20746479050569713, σ²⁻ = 0.0008764677906932903, σ²⁺ = 0.20698633696803714, σ³⁻ = 0.02933752665299981, σ³⁺ = 0.19221969947721135, σ⁴⁻ = 0.007153753618123909, σ⁴⁺ = 0.2355318734354983, d¹⁻ = 0.009678171103967854, d¹⁺ = 0.03307124754222082, d²⁻ = 0.0002304456260578315, d²⁺ = 0.034387625261472585, d³⁻ = 0.008151150038468974, d³⁺ = 0.03491762894753613, d⁴⁻ = 0.0012399987286105135, d⁴⁺ = 0.032143005033992526)

- $r_{\max} = 1.6$

    (σ¹⁻ = 0.09333104581659404, σ¹⁺ = 0.24635566838077005, σ²⁻ = 0.01434633634362714, σ²⁺ = 0.21519657245954776, σ³⁻ = 0.13588458685190982, σ³⁺ = 0.23516123233962588, σ⁴⁻ = 0.07865521103356753, σ⁴⁺ = 0.25711105016876556, d¹⁻ = 0.022976909523000744, d¹⁺ = 0.041093072721539926, d²⁻ = 0.0033498656600487757, d²⁺ = 0.03812069921249145, d³⁻ = 0.031143336144119814, d³⁺ = 0.04369032262640971, d⁴⁻ = 0.013926647436258096, d⁴⁺ = 0.03489141654920768)


### n = 6

- $r_{\max} = 0.1$

    (σ¹⁻ = 1.85282054099663e-5, σ¹⁺ = 0.023145966158814982, σ²⁻ = 1.3316905938470572e-16, σ²⁺ = 0.04545855634066476, σ³⁻ = 0.0, σ³⁺ = 0.043761137805796095, σ⁴⁻ = 0.0, σ⁴⁺ = 0.05714851411962113, d¹⁻ = 7.559811197279707e-6, d¹⁺ = 0.0029570261226166536, d²⁻ = 9.288659405715966e-17, d²⁺ = 0.0075490745109442646, d³⁻ = 0.0, d³⁺ = 0.009984345488074057, d⁴⁻ = 0.0, d⁴⁺ = 0.01101382121113853)

- $r_{\max} = 0.5$

    (σ¹⁻ = 0.000337835657510278, σ¹⁺ = 0.09611175443589484, σ²⁻ = 4.704649333061432e-17, σ²⁺ = 0.12935779651324714, σ³⁻ = 0.0, σ³⁺ = 0.09605328270859576, σ⁴⁻ = 0.0, σ⁴⁺ = 0.14126768387931046, d¹⁻ = 0.00016248267156440547, d¹⁺ = 0.012199158440778525, d²⁻ = 2.4286627566980857e-17, d²⁺ = 0.017904676999738228, d³⁻ = 0.0, d³⁺ = 0.016533319304101242, d⁴⁻ = 0.0, d⁴⁺ = 0.01759647999230751)

- $r_{\max} = 1.0$

    (σ¹⁻ = 0.004629168919222696, σ¹⁺ = 0.15686267464011588, σ²⁻ = 2.069816290044548e-5, σ²⁺ = 0.16870317737073243, σ³⁻ = 0.000521832379898929, σ³⁺ = 0.13817593805579154, σ⁴⁻ = 9.504285652145179e-5, σ⁴⁺ = 0.19419513063565386, d¹⁻ = 0.001761095247849796, d¹⁺ = 0.02069220387429978, d²⁻ = 2.1971583802098525e-6, d²⁺ = 0.02352702340271543, d³⁻ = 0.00010089845117082735, d³⁺ = 0.02094493801508495, d⁴⁻ = 3.2298218138080544e-6, d⁴⁺ = 0.022344940242406616)

- $r_{\max} = 1.6$

    (σ¹⁻ = 0.026697240026348098, σ¹⁺ = 0.1956370358732584, σ²⁻ = 0.00011394799840204368, σ²⁺ = 0.18126568279269073, σ³⁻ = 0.018880011833233207, σ³⁺ = 0.18011940649189523, σ⁴⁻ = 0.0026174952531703316, σ⁴⁺ = 0.2198428901428306, d¹⁻ = 0.007960949716567895, d¹⁺ = 0.02712022751062288, d²⁻ = 2.7387659546521067e-5, d²⁺ = 0.026707946290038847, d³⁻ = 0.004817543037722012, d³⁺ = 0.02674389875465382, d⁴⁻ = 0.0003886614203236218, d⁴⁺ = 0.025685030895763506)

### n = 7

- $r_{\max} = 0.1$

    (σ¹⁻ = 6.228219488172145e-6, σ¹⁺ = 0.017148739166719166, σ²⁻ = 8.836120243608663e-17, σ²⁺ = 0.02990942904221349, σ³⁻ = 0.0, σ³⁺ = 0.030051401586224245, σ⁴⁻ = 0.0, σ⁴⁺ = 0.03977488343893432, d¹⁻ = 2.3929953309858993e-6, d¹⁺ = 0.001906247772635547, d²⁻ = 6.255852080194834e-17, d²⁺ = 0.0037467587878987495, d³⁻ = 0.0, d³⁺ = 0.006084351099563782, d⁴⁻ = 0.0, d⁴⁺ = 0.006929644518177351)

- $r_{\max} = 0.5$

    (σ¹⁻ = 2.6370857468556757e-5, σ¹⁺ = 0.07412908611952779, σ²⁻ = 3.361876678227496e-17, σ²⁺ = 0.1011519704646714, σ³⁻ = 0.0, σ³⁺ = 0.08579369698350552, σ⁴⁻ = 0.0, σ⁴⁺ = 0.11464569933639311, d¹⁻ = 1.4941770271152993e-5, d¹⁺ = 0.00832643940739571, d²⁻ = 1.657232774358658e-17, d²⁺ = 0.011848598987561862, d³⁻ = 0.0, d³⁺ = 0.011803299298596671, d⁴⁻ = 0.0, d⁴⁺ = 0.012374097666620959)

- $r_{\max} = 1.0$

    (σ¹⁻ = 0.00036126863854391564, σ¹⁺ = 0.1239286677410143, σ²⁻ = 2.94083157596683e-17, σ²⁺ = 0.13941054732601338, σ³⁻ = 0.0, σ³⁺ = 0.12867546369158808, σ⁴⁻ = 0.0, σ⁴⁺ = 0.1643049357175383, d¹⁻ = 0.00016471918504498887, d¹⁺ = 0.014308283904097706, d²⁻ = 1.1247875676211357e-17, d²⁺ = 0.016597382823660155, d³⁻ = 0.0, d³⁺ = 0.016076398588076378, d⁴⁻ = 0.0, d⁴⁺ = 0.016322105591149976)

- $r_{\max} = 1.6$

    (σ¹⁻ = 0.004587703408332498, σ¹⁺ = 0.16168181943248916, σ²⁻ = 2.871658842842377e-17, σ²⁺ = 0.15615714170566458, σ³⁻ = 0.0003989927771879872, σ³⁺ = 0.16087348253904055, σ⁴⁻ = 0.0, σ⁴⁺ = 0.1932401059128731, d¹⁻ = 0.0016749053509084808, d¹⁺ = 0.0199617919452534, d²⁻ = 9.065236718721539e-18, d²⁺ = 0.02010094701129514, d³⁻ = 9.534845488988637e-5, d³⁺ = 0.019358896349746305, d⁴⁻ = 0.0, d⁴⁺ = 0.01968628506118533)

### n = 8

- $r_{\max} = 0.1$

    (σ¹⁻ = 0.0, σ¹⁺ = 0.014175689103804734, σ²⁻ = 6.452405827687369e-17, σ²⁺ = 0.024184033671799363, σ³⁻ = 0.0, σ³⁺ = 0.02129794088762056, σ⁴⁻ = 0.0, σ⁴⁺ = 0.03120834005743195, d¹⁻ = 0.0, d¹⁺ = 0.0014040485878980592, d²⁻ = 4.477656783734837e-17, d²⁺ = 0.002518797572488253, d³⁻ = 0.0, d³⁺ = 0.0037985056411898766, d⁴⁻ = 0.0, d⁴⁺ = 0.004603384301278449)

- $r_{\max} = 0.5$

    (σ¹⁻ = 1.6657983896418737e-6, σ¹⁺ = 0.06157728231332223, σ²⁻ = 2.607609205566357e-17, σ²⁺ = 0.08577355486022649, σ³⁻ = 0.0, σ³⁺ = 0.07014762632140267, σ⁴⁻ = 0.0, σ⁴⁺ = 0.09943735841090012, d¹⁻ = 9.070084904508329e-7, d¹⁺ = 0.006047375147467263, d²⁻ = 1.2270019739673873e-17, d²⁺ = 0.00845556984132353, d³⁻ = 0.0, d³⁺ = 0.008553992148170892, d⁴⁻ = 0.0, d⁴⁺ = 0.009250120293682237)

- $r_{\max} = 1.0$

    (σ¹⁻ = 3.0295521265476337e-5, σ¹⁺ = 0.10570148955546711, σ²⁻ = 1.118745181859527e-12, σ²⁺ = 0.12191782120532299, σ³⁻ = 0.0, σ³⁺ = 0.11097031650102797, σ⁴⁻ = 0.0, σ⁴⁺ = 0.14732335210133973, d¹⁻ = 1.2801004758255305e-5, d¹⁺ = 0.010882293320006748, d²⁻ = 1.9895316739650368e-12, d²⁺ = 0.012710358769407615, d³⁻ = 0.0, d³⁺ = 0.012271073835068654, d⁴⁻ = 0.0, d⁴⁺ = 0.012822202169077625)

- $r_{\max} = 1.6$

    (σ¹⁻ = 0.0005714021782451148, σ¹⁺ = 0.14061346859287138, σ²⁻ = 2.3848616065785948e-17, σ²⁺ = 0.1382538597870038, σ³⁻ = 0.0, σ³⁺ = 0.14489255793487546, σ⁴⁻ = 0.0, σ⁴⁺ = 0.17635385721794375, d¹⁻ = 0.00025871451057436073, d¹⁺ = 0.015586752896331959, d²⁻ = 7.043528535258468e-18, d²⁺ = 0.015896307872734546, d³⁻ = 0.0, d³⁺ = 0.015167293257391686, d⁴⁻ = 0.0, d⁴⁺ = 0.015460534725261878)


$n = 5$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0006|0.0|0.0|0.0|0.0357|0.0759|0.0676|0.0831|
|0.5|0.0046|0.0001|0.0006|0.0003|0.1373|0.1721|0.1367|0.1853|
|1.0|0.0307|0.0009|0.0293|0.0072|0.2075|0.2070|0.1922|0.2355|
|1.6|0.0933|0.0143|0.1359|0.0787|0.2464|0.2152|0.2352|0.2571|

|$r_{\max}$|$d^{1-}$|$d^{2-}$|$d^{3-}$|$d^{4-}$|$d^{1+}$|$d^{2+}$|$d^{3+}$|$d^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0003|0.0|0.0|0.0|0.0058|0.0155|0.0155|0.0169|
|0.5|0.0019|0.0|0.0001|0.0|0.0211|0.0281|0.0249|0.0267|
|1.0|0.0097|0.0002|0.0082|0.0012|0.0331|0.0344|0.0349|0.0321|
|1.6|0.0230|0.0034|0.0311|0.0139|0.0411|0.0381|0.0437|0.0349|

$n = 6$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0231|0.0455|0.0438|0.0571|
|0.5|0.0003|0.0000|0.0000|0.0000|0.0961|0.1294|0.0961|0.1413|
|1.0|0.0046|0.0000|0.0005|0.0001|0.1569|0.1687|0.1382|0.1942|
|1.6|0.0267|0.0001|0.0189|0.0026|0.1956|0.1813|0.1801|0.2198|


|$r_{\max}$|$d^{1-}$|$d^{2-}$|$d^{3-}$|$d^{4-}$|$d^{1+}$|$d^{2+}$|$d^{3+}$|$d^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0030|0.0075|0.0100|0.0110|
|0.5|0.0002|0.0000|0.0000|0.0000|0.0122|0.0179|0.0165|0.0176|
|1.0|0.0018|0.0000|0.0001|0.0000|0.0207|0.0235|0.0209|0.0223|
|1.6|0.0080|0.0000|0.0048|0.0004|0.0271|0.0267|0.0267|0.0257|

$n = 7$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0171|0.0299|0.0301|0.0398|
|0.5|0.0000|0.0000|0.0000|0.0000|0.0741|0.1012|0.0858|0.1146|
|1.0|0.0004|0.0000|0.0000|0.0000|0.1239|0.1394|0.1287|0.1643|
|1.6|0.0046|0.0000|0.0004|0.0000|0.1617|0.1562|0.1609|0.1932|

|$r_{\max}$|$d^{1-}$|$d^{2-}$|$d^{3-}$|$d^{4-}$|$d^{1+}$|$d^{2+}$|$d^{3+}$|$d^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0019|0.0037|0.0061|0.0069|
|0.5|0.0000|0.0000|0.0000|0.0000|0.0083|0.0118|0.0118|0.0124|
|1.0|0.0002|0.0000|0.0000|0.0000|0.0143|0.0166|0.0161|0.0163|
|1.6|0.0017|0.0000|0.0001|0.0000|0.0200|0.0201|0.0194|0.0197|

$n = 8$

|$r_{\max}$|$\sigma^{1-}$|$\sigma^{2-}$|$\sigma^{3-}$|$\sigma^{4-}$|$\sigma^{1+}$|$\sigma^{2+}$|$\sigma^{3+}$|$\sigma^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0142|0.0242|0.0213|0.0312|
|0.5|0.0000|0.0000|0.0000|0.0000|0.0616|0.0858|0.0701|0.0994|
|1.0|0.0000|0.0000|0.0000|0.0000|0.1057|0.1219|0.1110|0.1473|
|1.6|0.0006|0.0000|0.0000|0.0000|0.1406|0.1383|0.1449|0.1764|

|$r_{\max}$|$d^{1-}$|$d^{2-}$|$d^{3-}$|$d^{4-}$|$d^{1+}$|$d^{2+}$|$d^{3+}$|$d^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0014|0.0025|0.0038|0.0046|
|0.5|0.0000|0.0000|0.0000|0.0000|0.0060|0.0085|0.0086|0.0093|
|1.0|0.0000|0.0000|0.0000|0.0000|0.0109|0.0127|0.0123|0.0128|
|1.6|0.0003|0.0000|0.0000|0.0000|0.0156|0.0159|0.0152|0.0155|

### $n = 5$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 0.0012292655982896199, 𝓐¹⁺ = 0.07187403867569402, 𝓐²⁻ = 4.537924495082146e-16, 𝓐²⁺ = 0.1531243873253255, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.1366362230294785, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.16767902615045777)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 0.00958564623180895, 𝓐¹⁺ = 0.2758065125006352, 𝓐²⁻ = 3.7890027645501765e-5, 𝓐²⁺ = 0.34520356631972077, 𝓐³⁻ = 0.000764317217978492, 𝓐³⁺ = 0.2726270452104397, 𝓐⁴⁻ = 0.00018455351597496965, 𝓐⁴⁺ = 0.3718245965864604)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 0.06522011727308352, 𝓐¹⁺ = 0.4164834931865041, 𝓐²⁻ = 0.0032423626564446997, 𝓐²⁺ = 0.41562610584606047, 𝓐³⁻ = 0.06237200073595892, 𝓐³⁺ = 0.3892658658210833, 𝓐⁴⁻ = 0.019416684526763935, 𝓐⁴⁺ = 0.4726783093213293)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.19519911725524183, 𝓐¹⁺ = 0.4985628910940124, 𝓐²⁻ = 0.032161033885780614, 𝓐²⁺ = 0.43400317821650114, 𝓐³⁻ = 0.28108445061532666, 𝓐³⁺ = 0.47670995478819705, 𝓐⁴⁻ = 0.16505294873609058, 𝓐⁴⁺ = 0.5180635766985849)

### $n = 6$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 4.297648716842587e-5, 𝓐¹⁺ = 0.046126685329609514, 𝓐²⁻ = 2.996105757958167e-16, 𝓐²⁺ = 0.09022726467427147, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.08891544530332435, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.11387364832792198)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 0.0007304865858008984, 𝓐¹⁺ = 0.192879995115428, 𝓐²⁻ = 9.984050880687179e-17, 𝓐²⁺ = 0.2595572185671469, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.19310131605065742, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.28325901751602456)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 0.008924689138638656, 𝓐¹⁺ = 0.31318322049136726, 𝓐²⁻ = 8.156300685753815e-17, 𝓐²⁺ = 0.3358839765807297, 𝓐³⁻ = 0.0005282482411123644, 𝓐³⁺ = 0.27449677420767277, 𝓐⁴⁻ = 7.538653173355916e-5, 𝓐⁴⁺ = 0.3859175199060911)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.05455882398744686, 𝓐¹⁺ = 0.3937285380601262, 𝓐²⁻ = 0.00039816027686887906, 𝓐²⁺ = 0.36314326331831237, 𝓐³⁻ = 0.04176260016576256, 𝓐³⁺ = 0.3615421073216701, 𝓐⁴⁻ = 0.005291573769343571, 𝓐⁴⁺ = 0.439253710831755)

### $n = 6$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 7.67284716971498e-6, 𝓐¹⁺ = 0.03437594165630714, 𝓐²⁻ = 2.0204875936295446e-16, 𝓐²⁺ = 0.06056076192074657, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.06134265042343587, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.08055402297363452)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 4.032109383193416e-5, 𝓐¹⁺ = 0.14822204988070037, 𝓐²⁻ = 6.901567585207757e-17, 𝓐²⁺ = 0.2021183878742622, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.16920978391147404, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.22818904449013638)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 0.0009348718582917231, 𝓐¹⁺ = 0.2485196562632332, 𝓐²⁻ = 6.150134029685057e-17, 𝓐²⁺ = 0.2808287498506751, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2558601044149862, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.32891035790266226)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.00946220265836522, 𝓐¹⁺ = 0.32455376623547216, 𝓐²⁻ = 5.846893841942267e-17, 𝓐²⁺ = 0.31252492527004827, 𝓐³⁻ = 0.0006383137975704816, 𝓐³⁺ = 0.3222200160945192, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.3876220219032487)

### $n = 6$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 0.0, 𝓐¹⁺ = 0.028552343602762483, 𝓐²⁻ = 1.4275634583253184e-16, 𝓐²⁺ = 0.04827407644250388, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.042499953654163015, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.061603985384544396)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 1.243141551366449e-6, 𝓐¹⁺ = 0.1240488496952325, 𝓐²⁻ = 5.379823020375774e-17, 𝓐²⁺ = 0.17101934835738108, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.13893657246468463, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.19813452933092282)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 9.115277429253791e-5, 𝓐¹⁺ = 0.21169525044125143, 𝓐²⁻ = 4.8946302596376677e-17, 𝓐²⁺ = 0.24536746171032786, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2246172707967133, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.2962906961429553)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.0012802099046232485, 𝓐¹⁺ = 0.2827116687347921, 𝓐²⁻ = 4.8698573966670674e-17, 𝓐²⁺ = 0.2775177278262383, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2891726523037664, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.3531951275421345)